docs/testing/writing-unit-tests.md

```diff
@@ -0,0 +1,456 @@
+        const myModule = new MyModule();
+        expect(myModule.someFunc(input)).toBeTruthy();
+    });
+});
+```
+
+**3. Test async by returning Promises or async/await**
+
+It's best to use the `done` idiom that Jasmine provides only as a last resort. Using `async/await` or just returning a Promise allows Jasmine/Karma to just consume the test properly. Forcing the harness to wait on the `done` function to execute is, even according to the Jasmine docs, somewhat brittle and flaky. That said, there's plenty of cases where it can't be avoided, especially when testing code that uses a message bus, or fires jQuery events, or other things that work asynchronously, but don't return Promises. If you can't refactor your code to cough up a promise, then you're kinda stuck with `done`.
+
+**4. Try to be focused if stuck with using `done`**
+
+If you have a case that necessitates using `done` to finish an async test, be very clear on what you're testing. Here's an example from the BootstrapSearch module:
+```Javascript
+it('Should fire an input function when triggered by input', (done) => {
+    const bsSearch = new BootstrapSearch($targetElem, {
+        inputFunction: () => {
+            done();
+        }
+    });
+    bsSearch.val('stuff');
+});
+```
+This creates a new BootstrapSearch widget (`$targetElem` is set elsewhere) where its input function just returns done. This gets called by changing the input value to that search widget. We know the test passes if that gets invoked, so nothing else is needed. Then we trigger it by programmatically setting the value. That's it. No other code runs, no asynch searching calls, nothing else. Just validate that the given input function gets executed.
+
+Because of the async setup of this (there's no guarantee that `inputFunction` will get called immediately, that's up to the implementation), a simple spy won't work - how long should we wait to test whether it was invoked? It doesn't deal with Promises at all, so there's no thennable to handle or return. The best way to test this is with `done`.
+
+**5. Avoid using `setTimeout` if you can**
+
+It's an easy way to think of things - if you don't have access to a Promise, and code doesn't run in a way that you can link `done` to an event that gets fired, it would be straight forward to just run your code, then fire a timeout that runs `done` after some arbitrary time.
+
+The problem here is finding how long your timeout should be, and balancing that against the timeouts that Jasmine/Karma uses. These can then leave your module and other assorted variables in browser memory and scope a lot longer than other tests might expect! The timeout function can continue to linger after an `afterEach` function gets called, then when the timeout's up and your function gets run, some expected things might have been deleted. Worse, the test runner can move on at that point, and you'll see an error that looks like `"an error was thrown in afterAll!"`.
```

Comment:
Might be worth mentioning how you can specify the time limit for a test -- e.g. `it('does something slowly', () => { /* setup + test here */ }, 50000)`

Reply:
Huh, I didn't know you could do that! I'll add it.